src/main/java/io/cdap/plugin/batch/source/FTPBatchSource.java

```diff
@@ -317,5 +321,74 @@ private void validateAuthenticationInPath(FailureCollector collector) {
+     * and port number, then the username and password. Will timeout after 10 seconds
+     * if connection is not established.
+     *
+     * @param collector
+     */
+    boolean getConnection(FailureCollector collector) {
+      Path urlInfo;
+      String extractedPassword = extractPasswordFromUrl();
+      String encodedPassword = URLEncoder.encode(extractedPassword);
+      String validatePath = path.replace(extractedPassword, encodedPassword);
+      try {
+        urlInfo = new Path(validatePath);
+      } catch (Exception e) {
+        throw new IllegalArgumentException(String.format("Unable to parse url: %s %s", e.getMessage(), e));
+      }
+      // server is the substring between @ and : in the path
+      String server = urlInfo.toUri().getAuthority().substring(urlInfo.toUri().getAuthority().lastIndexOf("@") + 1,
+              urlInfo.toUri().getAuthority().lastIndexOf(":"));
+      String user = urlInfo.toUri().getAuthority().split(":")[0];
+      String protocol = urlInfo.toUri().getScheme();
+      int port = urlInfo.toUri().getPort();
+      if (port == -1 && protocol.equals(FTP_PROTOCOL)) {
+        port = DEFAULT_FTP_PORT;
+      }
+      if (port == -1 && protocol.equals(SFTP_PROTOCOL)) {
+        port = DEFAULT_SFTP_PORT;
+      }
+      if (protocol.equals(FTP_PROTOCOL)) {
+        return validateFTPConnection(collector, server, port, user, extractedPassword);
+      } else if (protocol.equals(SFTP_PROTOCOL)) {
+        return validateSFTPConnection(collector, server, port, user, extractedPassword);
+      }
+      return false;
+    }
+
+    private boolean validateFTPConnection(FailureCollector collector, String server, int port, String user, String password) {
+      boolean connected = false;
+      FTPClient ftpClient = new FTPClient();
+      // timeout after 5 seconds
+      ftpClient.setConnectTimeout(5000);
+      try {
+        ftpClient.connect(server, port);
+        try {
+          boolean isLogin = ftpClient.login(user, password);
+          if (!isLogin) {
+            collector.addFailure("Unable to authenticate with given username and password", null)
+                    .withConfigProperty(PATH);
+          } else {
+            connected = true;
+          }
+          ftpClient.disconnect();
+        } catch (Exception e) {
+          collector.addFailure("Unable to authenticate with given username and password", null)
+                  .withConfigProperty(PATH).withStacktrace(e.getStackTrace());
+        }
+      } catch (Exception e) {
+        collector.addFailure("Unable to establish connection with given host and port", null)
+                .withConfigProperty(PATH).withStacktrace(e.getStackTrace());
+      }
+      return connected;
+    }
+
+    private boolean validateSFTPConnection(FailureCollector collector, String server, int port, String user, String password) {
+      // TODO: add SFTP connection validation in future
```

Comment:
This is more about I'm not sure which method I should use for SFTP connection, found 3 online. I was wondering if the pipeline deployment part has codes that I can borrow since that will establish a connection